File: generate_route.py
import numpy as np

# Waypoints from gps_test_data.txt
waypoints = [
    (3.8480, 11.5020), # Emia
    (3.8490, 11.5030),
    (3.8510, 11.5050),
    (3.8528, 11.5067),
    (3.8548, 11.5088),
    (3.8558, 11.5098),
    (3.8580, 11.5120),
    (3.8590, 11.5130),
    (3.8594, 11.5134)  # Total Melen
]

def interpolate_points(waypoints, total_points=500):
    # Calculate total distance to normalize steps
    distances = []
    total_dist = 0
    for i in range(len(waypoints) - 1):
        p1 = np.array(waypoints[i])
        p2 = np.array(waypoints[i+1])
        dist = np.linalg.norm(p2 - p1)
        distances.append(dist)
        total_dist += dist
    
    # Generate points
    all_points = []
    
    # For each segment, calculate how many points it should have based on distance
    current_point_count = 0
    
    for i in range(len(waypoints) - 1):
        segment_dist = distances[i]
        # Ratio of total points for this segment
        segment_points_count = int((segment_dist / total_dist) * total_points)
        
        # Ensure at least 1 point if distance > 0
        if segment_points_count == 0 and segment_dist > 0:
            segment_points_count = 1
            
        p1 = np.array(waypoints[i])
        p2 = np.array(waypoints[i+1])
        
        # Linear interpolation
        for j in range(segment_points_count):
            t = j / segment_points_count
            point = p1 + t * (p2 - p1)
            all_points.append(tuple(point))
            
    # Add final point
    all_points.append(waypoints[-1])
    
    return all_points

points = interpolate_points(waypoints, 500)

# Output as Dart List
print("import 'package:latlong2/latlong.dart';")
print("")
print("final List<LatLng> emiaMelenRoute = [")
for p in points:
    print(f"  LatLng({p[0]:.6f}, {p[1]:.6f}),")
print("];")

File: Frontend/bin/generate_route.dart
import 'dart:convert';
import 'dart:io';
import 'dart:math' as math;

void main() async {
  // Total Melen -> Pharmacie Emia (REVERSED)
  final startLat = 3.8594;
  final startLng = 11.5134;
  final endLat = 3.8480;
  final endLng = 11.5020;

  final url =
      'http://router.project-osrm.org/route/v1/driving/$startLng,$startLat;$endLng,$endLat?overview=full&geometries=geojson';

  print('Fetching route from OSRM...');
  final request = await HttpClient().getUrl(Uri.parse(url));
  final response = await request.close();

  if (response.statusCode != 200) {
    print('Error fetching route: ${response.statusCode}');
    exit(1);
  }

  final responseBody = await response.transform(utf8.decoder).join();
  final data = jsonDecode(responseBody);

  if (data['routes'] == null || (data['routes'] as List).isEmpty) {
    print('No route found');
    exit(1);
  }

  final geometry = data['routes'][0]['geometry'];
  final coordinates = geometry['coordinates'] as List;

  // Convert [lng, lat] to [lat, lng]
  final List<List<double>> waypoints = coordinates.map((c) {
    final point = c as List;
    return [point[1] as double, point[0] as double];
  }).toList();

  print('Route fetched with ${waypoints.length} waypoints.');

  // Interpolate to 500 points
  final interpolatedPoints = interpolatePoints(waypoints, 500);

  print('Interpolated to ${interpolatedPoints.length} points.');

  // Generate Dart file
  final buffer = StringBuffer();
  buffer.writeln("import 'package:latlong2/latlong.dart';");
  buffer.writeln("");
  buffer.writeln("// Route: Total Melen -> Pharmacie Emia");
  buffer.writeln("// Generated from OSRM");
  buffer.writeln("final List<LatLng> emiaMelenRoute = [");

  for (final point in interpolatedPoints) {
    buffer.writeln("  LatLng(${point[0]}, ${point[1]}),");
  }

  buffer.writeln("];");

  final file = File('lib/utils/simulation_data.dart');
  await file.writeAsString(buffer.toString());

  print('File generated at lib/utils/simulation_data.dart');
}

List<List<double>> interpolatePoints(
  List<List<double>> waypoints,
  int totalPoints,
) {
  final distances = <double>[];
  double totalDist = 0;

  for (int i = 0; i < waypoints.length - 1; i++) {
    final d = distance(waypoints[i], waypoints[i + 1]);
    distances.add(d);
    totalDist += d;
  }

  final allPoints = <List<double>>[];

  for (int i = 0; i < waypoints.length - 1; i++) {
    final segmentDist = distances[i];
    int segmentCount = ((segmentDist / totalDist) * totalPoints).round();

    if (segmentCount < 1) segmentCount = 1;

    final p1 = waypoints[i];
    final p2 = waypoints[i + 1];

    for (int j = 0; j < segmentCount; j++) {
      final t = j / segmentCount;
      final lat = p1[0] + t * (p2[0] - p1[0]);
      final lng = p1[1] + t * (p2[1] - p1[1]);
      allPoints.add([lat, lng]);
    }
  }

  // Add last point
  allPoints.add(waypoints.last);

  return allPoints;
}

double distance(List<double> p1, List<double> p2) {
  return math.sqrt(math.pow(p2[0] - p1[0], 2) + math.pow(p2[1] - p1[1], 2));
}
